### PR TITLE
HSEARCH-4065 Expose Lucene's low-level IndexReader through API

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-4065

### CODE REVIEW

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/impl/LuceneIndexScope.java [hunk 1]
```diff
@@ -19,9 +24,12 @@
 import org.hibernate.search.engine.backend.scope.spi.IndexScope;
 import org.hibernate.search.engine.search.aggregation.spi.SearchAggregationBuilderFactory;
 
+import org.apache.lucene.index.IndexReader;
+
 
 public class LuceneIndexScope
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/impl/LuceneIndexScope.java:30]: As Sonar suggested, you should rename this class to `LuceneIndexScopeImpl`.

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.Collections;
+import java.util.Set;
+
+import org.apache.lucene.index.IndexReader;
+
+/**
+ * The {@code LuceneIndexScope} exposes {@link IndexReader}s directly, making it possible to query the Lucene
+ * indexes directly bypassing Hibernate Search.
+ * <p>
+ * The returned IndexReader instances are always read-only and must be closed
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java:18]: I'm not sure why an index reader would *not* be read-only. That must be referring to some dodgy feature that was removed from Lucene.

```suggestion
 * The returned IndexReader instances must be closed
```

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	/**
+	 * Opens an IndexReader on all indexes containing the entities of the search scope.
+	 * In the simplest case if the scope has one entity, it will map to a single index; if the entity
+	 * uses a sharding strategy or if multiple entities using different index names are selected,
+	 * the single IndexReader will act as a MultiReader on the aggregate of these indexes.
+	 * This MultiReader is not filtered by Hibernate Search, so it might contain information
+	 * relevant to different types as well.
+	 * <p>
+	 * The returned IndexReader is read only; writing directly to the index is discouraged.
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java:35]: Let's keep this simple:

* We can just state that the reader targets all indexes in this scope, that will cover the explanation of what happens if there is a single entity or multiple entities.
* Hibernate Search 6 stores at most one type per index, so the reader cannot "contain information relevant to different types as well".
* As explained above, an index reader is obviously read-only.

```suggestion
	 * Opens an IndexReader for all indexes containing the entities of the search scope.
```

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	 * <p>
+	 * The instance must be closed after its use by the client of this class using {@link IndexReader#close()}.
+	 * <p>
+	 * By default, if routing is not configured, all shards will be queried.
+	 * If you need to filter by routing keys use {@link #openIndexReader(Set)}.
+	 * This api bypasses the filtering by tenantId, so results from all tenants will be always visible.
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java:41]:
```suggestion
	 * If sharding is enabled, the returned reader will read all shards.
	 * If you need to read specific shards only, use {@link #openIndexReader(Set)}.
	 * <p>
	 * <strong>WARNING:</strong> Even if multi-tenancy is enabled, the returned reader exposes documents of *all* tenants.
```

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+	/**
+	 * Opens an IndexReader on all indexes containing the entities of the search scope.
+	 * In the simplest case if the scope has one entity, it will map to a single index; if the entity
+	 * uses a sharding strategy or if multiple entities using different index names are selected,
+	 * the single IndexReader will act as a MultiReader on the aggregate of these indexes.
+	 * This MultiReader is not filtered by Hibernate Search, so it might contain information
+	 * relevant to different types as well.
+	 * <p>
+	 * The returned IndexReader is read only; writing directly to the index is discouraged.
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java:57]: Let's keep this simple:

* We can just state that the reader targets all indexes in this scope, that will cover the explanation of what happens if there is a single entity or multiple entities.
* Hibernate Search 6 stores at most one type per index, so the reader cannot "contain information relevant to different types as well".
* As explained above, an index reader is obviously read-only.

```suggestion
	 * Opens an IndexReader on shards assigned to the given routing keys
	 * for all indexes containing the entities of the search scope.
```

File: backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	 * <p>
+	 * The instance must be closed after its use by the client of this class using {@link IndexReader#close()}.
+	 * <p>
+	 * Only the shards having keys contained in {@code routingKeys} will be visible to the reader.
+	 * This api bypasses the filtering by tenantId, so results from all tenants will be always visible.
```
Comment (yrodiere) [inline backend/lucene/src/main/java/org/hibernate/search/backend/lucene/scope/LuceneIndexScope.java:62]:
```suggestion
	 * <strong>WARNING:</strong> Even if multi-tenancy is enabled, the returned reader exposes documents of *all* tenants.
```

File: documentation/src/main/asciidoc/reference/backend-lucene.asciidoc [hunk 7]
```diff
@@ -1217,3 +1217,34 @@ include::{sourcedir}/org/hibernate/search/documentation/backend/lucene/indexmana
 <4> Compute the index size and get the result.
 <5> An asynchronous version of the method is also available.
 ====
+
+== Retrieving low-level Lucene `IndexReader`
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-lucene.asciidoc:1221]:
```suggestion
== Retrieving a Lucene `IndexReader`
```

File: documentation/src/main/asciidoc/reference/backend-lucene.asciidoc [hunk 8]
```diff
@@ -1217,3 +1217,34 @@ include::{sourcedir}/org/hibernate/search/documentation/backend/lucene/indexmana
...
+.Retrieving the index reader from a Lucene index scope
+====
+[source, JAVA, indent=0, subs="+callouts"]
+----
+include::{sourcedir}/org/hibernate/search/documentation/backend/lucene/lowlevel/LuceneLowLevelIndexReaderIT.java[tags=indexReader]
+----
+<1> Retrieve the `SearchMapping`.
+<2> Retrieve the `LuceneIndexScope` extending the search scope with the `LuceneExtension`.
+<3> Open an instance of `IndexReader`. Optionally it is possible to provide the routing keys
+to filter the visibility of the returning index reader using the method `openIndexReader(Set<String>)`.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-lucene.asciidoc:1234]:
```suggestion
to target only some shards of the indexes using the method `openIndexReader(Set<String>)`.
```

File: documentation/src/main/asciidoc/reference/backend-lucene.asciidoc [hunk 9]
```diff
@@ -1217,3 +1217,34 @@ include::{sourcedir}/org/hibernate/search/documentation/backend/lucene/indexmana
...
+====
+
+[NOTE]
+====
+The returned IndexReader is read only; writing directly to the index is discouraged.
+====
+
+[IMPORTANT]
+====
+The returned IndexReader instance must be closed after its use by the client.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-lucene.asciidoc:1244]: I would move this to the example. Wrap the use of the index reader in a try-with-resource block and add a callout that says "The `IndexReader` must be closed after use".

File: documentation/src/main/asciidoc/reference/backend-lucene.asciidoc [hunk 10]
```diff
@@ -1217,3 +1217,34 @@ include::{sourcedir}/org/hibernate/search/documentation/backend/lucene/indexmana
...
+<2> Retrieve the `LuceneIndexScope` extending the search scope with the `LuceneExtension`.
+<3> Open an instance of `IndexReader`. Optionally it is possible to provide the routing keys
+to filter the visibility of the returning index reader using the method `openIndexReader(Set<String>)`.
+====
+
+[NOTE]
+====
+The returned IndexReader is read only; writing directly to the index is discouraged.
+====
+
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-lucene.asciidoc:1241]:
```suggestion
```

As stated above, an index reader is obviously read-only. This note is not necessary.

File: documentation/src/main/asciidoc/reference/backend-lucene.asciidoc [hunk 11]
```diff
@@ -1217,3 +1217,34 @@ include::{sourcedir}/org/hibernate/search/documentation/backend/lucene/indexmana
...
+[IMPORTANT]
+====
+The returned IndexReader instance must be closed after its use by the client.
+====
+
+[WARNING]
+====
+This api bypasses the filtering by tenantId, so results from all tenants will be always visible.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-lucene.asciidoc:1249]:
```suggestion
Even if multi-tenancy is enabled, the returned reader exposes documents of *all* tenants.
```

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/scope/ScopeExtensionIT.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+	public BackendMock backendMock = new BackendMock();
+
+	@Rule
+	public OrmSetupHelper ormSetupHelper = OrmSetupHelper.withBackendMock( backendMock );
+
+	private SessionFactory sessionFactory;
+
+	@Before
+	public void setup() {
+		backendMock.expectAnySchema( SearchQueryBaseIT.Book.NAME );
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/scope/ScopeExtensionIT.java:36]: Please do not reuse entities from another test.

File: integrationtest/mapper/pojo-base/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/scope/ScopeExtensionIT.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+	@Rule
+	public JavaBeanMappingSetupHelper setupHelper =
+			JavaBeanMappingSetupHelper.withBackendMock( MethodHandles.lookup(), backendMock );
+
+	private SearchMapping mapping;
+
+	@Before
+	public void setup() {
+		backendMock.expectAnySchema( BasicIndexedEntity.NAME );
```
Comment (yrodiere) [inline integrationtest/mapper/pojo-base/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/scope/ScopeExtensionIT.java:37]: Please do not reuse entities from another test.

**Review by yrodiere — COMMENTED** [review]
A few comments, especially around documentation. Feel free to merge once they are addressed. Thanks!

File: documentation/src/main/asciidoc/reference/backend-lucene.asciidoc (hunk 9 above)
Comment (fax4ever) [reply documentation/src/main/asciidoc/reference/backend-lucene.asciidoc:1239]: I've already wrapped the use of the index reader in a try-with-resource block. Then I'm just adding the callout.

File: integrationtest/mapper/pojo-base/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/scope/ScopeExtensionIT.java (hunk 13 above)
Comment (fax4ever) [reply integrationtest/mapper/pojo-base/src/test/java/org/hibernate/search/integrationtest/mapper/pojo/scope/ScopeExtensionIT.java:37]: I think that this entity is already used from different tests. Anyway I'm going to create a new entity for this test.

**Comment by fax4ever** [comment]
Thanks @yrodiere, I applied all the suggestions here.
I'll merge it when the CI passes.

**Pull request merged**